rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs

```diff
@@ -0,0 +1,230 @@
+        minimum_throughput: Throughput,
+        time_source: SharedTimeSource,
+        async_sleep: SharedAsyncSleep,
+    }
+}
+
+impl<T> MinimumThroughputBody<T> {
+    // If this function returns:
+    // - `None`: We couldn't calculate a throughput because we don't have good data yet.
+    // - `Some(throughput)`: We have good data, and we can calculate a throughput.
+    // -- `Err(_)`: A bug occurred.
+    fn calculate_throughput(&self) -> Result<Option<Throughput>, Error> {
+        if let Some((earliest_time, _)) = self.throughput_logs.front() {
+            let now = self.time_source.now();
+            let time_elapsed_since_earliest_poll = now
+                .duration_since(*earliest_time)
+                .map_err(|err| Error::TimeTravel(err.into()))?;
+
+            // This check ensures we that the data we're looking at covers a good range of time.
+            // If not, then we don't calculate a throughput.
+            if time_elapsed_since_earliest_poll < self.minimum_throughput.per_time_elapsed {
+                return Ok(None);
+            }
+
+            let total_bytes_logged =
+                self.throughput_logs
+                    .iter()
+                    .fold(0, |acc, (_, bytes_read)| acc + bytes_read) as f64;
+
+            Ok(Some(Throughput {
+                bytes_read: total_bytes_logged,
+                per_time_elapsed: time_elapsed_since_earliest_poll,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl<T: http_body::Body> MinimumThroughputBody<T> {
+    /// Given an HTTP body and a minimum throughput, create a new `MinimumThroughputBody`.
+    pub fn new(
+        time_source: impl TimeSource + 'static,
+        async_sleep: impl AsyncSleep + 'static,
+        body: T,
+        minimum_throughput: (u64, Duration),
+    ) -> Self {
+        Self {
+            inner: body,
+            throughput_logs: VecDeque::with_capacity(LOG_WINDOW_SIZE),
+            minimum_throughput: minimum_throughput.into(),
+            time_source: time_source.into_shared(),
+            async_sleep: async_sleep.into_shared(),
+        }
+    }
+}
+
+impl<T> http_body::Body for MinimumThroughputBody<T>
+where
+    T: http_body::Body<Data = bytes::Bytes, Error = Box<dyn std::error::Error + Send + Sync>>,
+{
+    type Data = T::Data;
+    type Error = T::Error;
+
+    fn poll_data(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        // It's possible that this body won't be polled again, so we ensure it'll
+        // wake by spawning a task with a reference to the waker, forcing another
+        // poll.
+        let wake_after = self.minimum_throughput.per_time_elapsed;
+        let waker = cx.waker().clone();
+        let sleeper = self.async_sleep.clone();
+        let _task = tokio::task::spawn(async move {
+            sleeper.sleep(wake_after).await;
+            waker.wake();
+        });
```

Comment:
Seems like this task shouldn't get spawned if we're going to return `Poll::ready`. Maybe split out most of this function into a separate function, and then here, call that function and examine the poll result before spawning this task.